Dockerfile

```diff
@@ -3,47 +3,60 @@
 ####################################################################################################
 ################################## Stage: builder ##################################################
 
-FROM balenalib/raspberry-pi-debian-python:buster-run-20211014 as builder
-
-ENV PYTHON_DEPENDENCIES_DIR=/opt/python-dependencies
-
-RUN mkdir /tmp/build
-COPY ./ /tmp/build
-WORKDIR /tmp/build
-
+FROM balenalib/raspberry-pi-debian:buster-build-20211014 as builder
 
+# Nebra uses /opt by convention
+WORKDIR /opt/
 
+# this installs python 3.7.3 which is buster default and supported by grpcio wheel
 RUN \
     install_packages \
-            build-essential \
-            libdbus-glib-1-dev
-
-RUN pip3 install --no-cache-dir --target="$PYTHON_DEPENDENCIES_DIR" .
+        python3-minimal \
+        python3-venv \
+        python3-pip \
+        build-essential \
+        libdbus-glib-1-dev
+
+# This will be the path that venv uses for installation below
+ENV PATH="/opt/venv/bin:$PATH"
+
+# updating pip to recent version otherwise it doesn't find correct grpcio
+# without pinning the pip version, we get docker linter warning
+COPY requirements.txt requirements.txt
+RUN python3 -m venv /opt/venv && \
+    pip install --no-cache-dir --upgrade pip==22.0.1 && \
+    pip install --no-cache-dir -r requirements.txt
 
 # firehose build, the tar is obtained from  quectel.
 # there is no install target in Makefile, doing manual copy
-RUN tar -xf quectel/qfirehose/QFirehose_Linux_Android_V1.4.9.tar
+COPY quectel quectel
+RUN tar -xf ./quectel/qfirehose/QFirehose_Linux_Android_V1.4.9.tar
 # docker linter wants WORKDIR for changing directory
-WORKDIR /tmp/build/QFirehose_Linux_Android_V1.4.9
+WORKDIR /opt/QFirehose_Linux_Android_V1.4.9
 RUN make && \
     cp QFirehose /usr/sbin/QFirehose && \
     rm -rf quectel/qfirehose
 
+WORKDIR /opt
+COPY ./ /opt
+RUN pip --no-cache-dir install .
+
 # No need to cleanup the builder
 
 ####################################################################################################
 ################################### Stage: runner ##################################################
 
-FROM balenalib/raspberry-pi-debian-python:buster-run-20211014 as runner
-
-ENV PYTHON_DEPENDENCIES_DIR=/opt/python-dependencies
+FROM balenalib/raspberry-pi-debian:buster-build-20211014 as runner
```

Comment:
No, don't do this. This container is *a lot* larger.

Reply:
> No, don't do this. This container is _a lot_ larger.

I can probably get rid of virtual env, but using balena's python creates problem with compiled wheel of grpcio.

Reply:
We've had the conversation of  venv vs no venv. At the end of the day it doesn't really matter if you copy over `site-packages` or the `venv` environment from a size perspective.

With regards to the compilation -- i don't see why that matters if we cache the previous layer. It will just be skipped.

Reply:
I want to keep the `balenalib/raspberry-pi-debian-python:buster-run-20211014 as runner`, since this is shared across all other containers (and thus won't add any size for the base layer).

I'm not talking about pip caching, i'm talking about docker caching. You can read more [here](https://dev.to/dtinth/caching-docker-builds-in-github-actions-which-approach-is-the-fastest-a-research-18ei).

Reply:
hm-config installs its own python, that can switch to balenalib/raspberry-pi-debian
hm-pktforwarder can also use the same. Doesn't have any specific dependency.

caching: I was also talking about docker caching, it is not working for source installed pip packages. Pip always compiles them and invalidate the layer cache.

Reply:
That really shouldn't be the case. It should only invalidate when `requirements.txt` is altered 

https://stackoverflow.com/questions/48441296/pip-compiled-as-part-of-dockerfile-fastest-way-to-add-a-new-entry-to-requireme

Reply:
the last "pip install . " which install hm-diag itself will mostly change because one of its source files would have changed. And then it will lookup dependencies -> which will be pyhelper -> which will depend on grpcio and that can only be installed with python setup.py because no wheel will be found matching python 3.10 and glibc-1.28. And we will be back to waiting 1hr 40 mins. 

There might be a way to preserve this cache. But each iteration I try on this will cost 1.4 hours. switching other to use balenalib/raspberry-pi-debian might be simpler solution IMHO. balenalib/raspberry-pi-debian-python will anyway likely be just one more layer on raspberry-pi-debian even if we don't switch.

Reply:
Sure, but that is easily solved using a simple `requirements.txt` in conjunction with `setup.py` dependencies. That way by the time it gets to `pip install .`, it will already have all the dependencies installed from a previous layer where `requirements.txt` was installed.

Reply:
that is already done in this pull request. The dockerfile was written to ensure that cache invalidation is a bare minimum. 

If using balenalib/raspberry-pi-debian-python instead of  balenalib/raspberry-pi-debian, that is acceptable. I don't think it will make a difference apart from having two python versions in the final container. But I am ok with that.

Reply:
I have started using balena-debian-python again. Though I think if we have use a python image, we should push our own based on balena-debian.